crates/ark/src/repos.rs

```diff
@@ -0,0 +1,298 @@
+fn find_repos_conf() -> Option<PathBuf> {
+    if let Some(path) = find_repos_conf_xdg("rstudio".to_string()) {
+        return Some(path);
+    }
+    if let Some(path) = find_repos_conf_xdg("ark".to_string()) {
+        return Some(path);
+    }
+    None
+}
+
+/// Apply the given default repositories to the R session.
+fn apply_repos(repos: HashMap<String, String>) -> anyhow::Result<()> {
+    log::debug!("Applying default repositories: {:?}", repos);
+    // Convert the HashMap to an R named character vector
+    let named_repos = RObject::from(repos);
+
+    // Call `apply_repo_defaults` to set the repos on the R side
+    let mut call = RFunction::new("", "apply_repo_defaults");
+    call.add(named_repos);
+    match call.call_in(ARK_ENVS.positron_ns) {
+        Ok(_) => Ok(()),
+        Err(e) => Err(anyhow::anyhow!(
+            "Error applying default repositories: {}",
+            e
+        )),
+    }
+}
+
+/// Apply the repos configuration file at the given path.
+///
+/// The repos configuration file is a simple INI-style configuration file styled after RStudio's
+/// /etc/rstudio/repos.conf. It is expected to consist of repository names and URLs, one per line,
+/// in the format `name=url`. Empty lines or lines beginning with `#` (comments) are ignored.
+///
+/// Arguments:
+/// - `path`: The path to the repos configuration file.
+///
+/// Returns:
+///
+/// `Ok(())` if the file was successfully read and applied, or an error if there was a
+/// problem.
+pub fn apply_repos_conf(path: PathBuf) -> anyhow::Result<()> {
+    log::info!("Using repos file at {:?}", path);
+    // Read the repos file
+    let file = std::fs::File::open(&path)?;
+    let reader = std::io::BufReader::new(file);
+    let mut repos = std::collections::HashMap::<String, String>::new();
+    for line in reader.lines() {
+        let line = line?;
+        // Ignore the line if it's only whitespace or starts with a comment
+        if line.trim().is_empty() || line.trim().starts_with('#') {
+            continue;
+        }
+        let parts: Vec<&str> = line.split('=').collect();
+        if parts.len() != 2 {
+            log::debug!("Skipping invalid line in repos file: {}", line);
+            continue;
+        }
+        let repo_name = parts[0].trim();
+        let repo_url = parts[1].trim();
+        repos.insert(repo_name.to_string(), repo_url.to_string());
+    }
+    apply_repos(repos)
+}
+
+/// Checks the Linux distribution name and version to determine the appropriate P3M repository URL.
+#[cfg(target_os = "linux")]
+fn get_p3m_linux_repo(linux_name: String) -> String {
+    // The following Linux names have 1:1 mappings to a P3M repository URL
+    let repo_names = [
+        String::from("bookworm"),
+        String::from("bullseye"),
+        String::from("focal"),
+        String::from("jammy"),
+        String::from("noble"),
+        String::from("opensuse155"),
+        String::from("opensuse156"),
+        String::from("rhel9"),
```

Comment:
Yes, it will need to be updated regularly. I ported this code from an identical list in `rstudio/rstudio-pro` which is also hardcoded ... but in C++. Maybe a config file or script would be best.